src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,99 @@
 package redis.clients.jedis;
 
 import java.nio.ByteBuffer;
-import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-
-import redis.clients.jedis.exceptions.JedisException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
 import redis.clients.jedis.util.SafeEncoder;
 
-public class ClientSideCache {
+public abstract class ClientSideCache {
 
-  private final Map<ByteBuffer, Object> cache;
+  private final Map<ByteBuffer, Set<Long>> keyHashes;
+  private final ReentrantLock writeLock = new ReentrantLock();
 
-  public ClientSideCache() {
-    this.cache = new HashMap<>();
+  protected ClientSideCache() {
+    this.keyHashes = new ConcurrentHashMap<>();
   }
 
-  /**
-   * For testing purpose only.
-   * @param map 
-   */
-  ClientSideCache(Map<ByteBuffer, Object> map) {
-    this.cache = map;
-  }
+  public abstract void invalidateAll();
 
-  public final void clear() {
-    cache.clear();
-  }
+  protected abstract void invalidateAll(Iterable<Long> hashes);
 
-  public final void invalidateKeys(List list) {
+  final void invalidate(List list) {
```

Comment:
There is a lot of invalidation going on in this class. I preferred the original name for this method, i.e. `invalidateKeys`, because it points out what is being invalidated. Along the same lines, the method `invalidateAll(Iterable<Long>)` could be named `invalidateHashes`. I'm not sure about `invalidateAll()`, maybe it could be `invalidateAllHashes()`? `invalidate0` could stay `invalidateKey`.

You know the joke with the two difficult problems in computer science: cache invalidation and naming/

Reply:
@gerzse This `invalidate(List)` is called only when the server pushes with the header `invalidate`. So it seemed fit.

As for `invalidateAll()` and `invalidateAll(Iterable)`, the names are aligned with the respective methods in both Guava and Caffeine libraries.

Reply:
some renaming are done. please re-check